File: app/lib/llm/schema.ts
import { z } from "zod";

export const InputSchema = z
  .object({
    id: z.string({ required_error: "ID is required" }).min(1, "ID is required"),
    text: z
      .string({ required_error: "Text is required" })
      .min(1, "Text is required"),
  })
  .strict();

export const EmotionSchema = z
  .object({
    neutral: z
      .number({
        required_error: "Neutral score is required",
      })
      .min(0, "Joy value must be 0 or greater")
      .max(1, "Joy value must be 1 or less"),
    happy: z
      .number({
        required_error: "Happy score is required",
      })
      .min(0, "Fun value must be 0 or greater")
      .max(1, "Fun value must be 1 or less"),
    sad: z
      .number({
        required_error: "Sad score is required",
      })
      .min(0, "Sorrow value must be 0 or greater")
      .max(1, "Sorrow value must be 1 or less"),
    angry: z
      .number({
        required_error: "Angry score is required",
      })
      .min(0, "Angry value must be 0 or greater")
      .max(1, "Angry value must be 1 or less"),
  })
  .strict();

export const ResponseSchema = z
  .object({
    role: z
      .string({ required_error: "Role is required" })
      .min(1, "Role is required"),
    content: z
      .string({ required_error: "Content is required" })
      .min(1, "Content is required"),
    emotion: EmotionSchema,
    event: z.union([
      z.literal("sit"),
      z.literal("go_to_user_position"),
      z.null(),
    ]),
    voice: z
      .string({ required_error: "Voice is required" })
      .base64()
      .min(1, "Voice is required"),
  })
  .strict();

File: ai-assistant/src/schema.py
from typing import Literal
from pydantic import BaseModel, Field


class Emotion(BaseModel):
    neutral: float = Field(description="ニュートラルの感情値(0〜1の間)")
    happy: float = Field(description="幸せの感情値(0〜1の間)")
    sad: float = Field(description="悲しみの感情値(0〜1の間)")
    angry: float = Field(description="怒りの感情値(0〜1の間)")


class EmotionMessage(BaseModel):
    role: str = Field(description="メッセージの役割(assistant)")
    content: str = Field(description="メッセージ内容")
    emotion: Emotion = Field(description="感情オブジェクト")
    event: Literal["sit", "go_to_user_position"] | None = Field(
        None, description="キャラクターの行動を示すテキスト"
    )
    game_ai_choice: Literal["rock", "paper", "scissors"] | None = Field(
        None, description="AIのじゃんけんの選択"
    )
    game_result: Literal["win", "lose", "draw"] | None = Field(
        None, description="じゃんけんの結果（ユーザー視点）"
    )


class Response(BaseModel):
    role: str = Field(description="メッセージの役割(assistant)")
    content: str = Field(description="メッセージ内容")
    emotion: Emotion = Field(description="感情オブジェクト")
    event: Literal["sit", "go_to_user_position"] | None = Field(
        None, description="キャラクターの行動を示すテキスト"
    )
    voice: str = Field(description="Base64エンコードされた音声データ")
    game_ai_choice: Literal["rock", "paper", "scissors"] | None = Field(
        None, description="AIのじゃんけんの選択"
    )
    game_result: Literal["win", "lose", "draw"] | None = Field(
        None, description="じゃんけんの結果（ユーザー視点）"
    )
